Respect the [[clang::unsafe_buffer_usage]] attribute for field and constructor initializers

danakj opened the pull request:
This is built on top of https://github.com/llvm/llvm-project/pull/91777, and includes the commits there for now.

Issue #80482

danakj committed:
Generate -Wunsafe-buffer-usage warnings in ctor and field initializers

Field initializers must be found by visiting FieldDecl and then
running the warning checks/matchers against the in-class initializer,
which is itself a Stmt. This catches warnings in places such as:

struct C {
  P* Ptr;
  AnUnsafeCtor U{Ptr};
};

CXXCtorInitializers are not statements either, but they point to an
initializer expression which is. When visiting a FunctionDecl, also walk
through any constructor initializers and run the warning checks/matchers
against their initializer expressions. This catches warnings for
initializing fields and calling other constructors, such as:

struct C {
  C(P* Ptr) : AnUnsafeCtor(Ptr) {}
}

We add tests for explicit construction, for field initialization, base
class constructor calls, delegated constructor calls, and aggregate
initialization.

Note that aggregate initialization through `()` is treated differently
today by the AST matchers than `{}`. The former is not considered as
calling an implicit constructor, while the latter is.
MatchDescendantVisitor::shouldVisitImplicitCode() returns false with a
TODO, which means we do not catch warnings of the form:

struct AggregateInitType { AnUnsafeCtor U; }
AggregateInitType{Ptr};

But we do already catch them when written as (in C++20):

struct AggregateInitType { AnUnsafeCtor U; }
AggregateInitType(Ptr);

Returning true from MatchDescendantVisitor::shouldVisitImplicitCode(),
however, breaks expectations for field in-class initializers by moving
the SourceLocation, possibly to inside the implicit ctor instead of on
the line where the field initialization happens.

struct C {
  P* Ptr;
  AnUnsafeCtor U{Ptr};  // expected-warning{{this is never seen then}}
};

Tests are also added for std::span(ptr, size) constructor being called
from a field initializer and a constructor initializer.

danakj commented:
These lit test failures are indeed from this PR's last commit, I will dig into why and correct what's wrong.

haoNoQ reviewed (commented):
Aha makes sense! Thanks for catching these false negatives!!

haoNoQ commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -3328,3 +3300,63 @@ void clang::checkUnsafeBufferUsage(const Decl *D,
+    }
+  }
+
+  if (const auto *FD = dyn_cast<FieldDecl>(D)) {
```
Actually, do we even need to handle default field initializers explicitly? I think the AST copy-pastes them into `CXXCtorInitializer`s in every constructor with the help of `CXXDefaultInitExpr`. So as long as you handle `CXXCtorInitializer` you get coverage of default initializers too. Unless there are zero constructors that respect that initializer. But then it's just dead code so we might as well leave it be.

danakj replied:
Unfortunately, CXXCtorInitializers do not catch everything. This test fails:

```cc
struct S {
    S()
        : FromCtor(3),  // expected-warning{{function introduces unsafe buffer manipulation}}
          FromCtor2{3}  // expected-warning{{function introduces unsafe buffer manipulation}}
    {}

    UnsafeMembers FromCtor;
    UnsafeMembers FromCtor2;
    UnsafeMembers FromField{3};  // expected-warning{{function introduces unsafe buffer manipulation}}
};
```

Where the int ctor of UnsafeMembers is `[[clang::unsafe_buffer_usage]]`.

This also fails:
```
struct AggregateUnsafeMembers {
    UnsafeMembers f1;
    UnsafeMembers f2{3};  // expected-warning{{function introduces unsafe buffer manipulation}}
};
```

haoNoQ replied:
Ok I debugged this a little bit and it looks like you're blaming the wrong visitor. The `CXXCtorInitializer` is there and it gets correctly visited. However, `MatchDescendantVisitor` gets stuck because it considers `CXXDefaultInitExpr` implicit code. Enabling `MatchDescendantVisitor::shouldVisitImplicitCode()` takes care of your issue!

Not sure it's a good solution though, might be a good idea to add an explicit `TraverseCXXDefaultInitExpr()` because I'm not sure we want to visit other kinds of implicit code. (Probably deserves a FIXME.)

danakj replied:
Thanks! I will have a look at that on Monday then.

danakj replied:
Hm, I changed `MatchDescendantVisitor::shouldVisitImplicitCode()` to return true, but that is not sufficient to resolve things for me. Did you try running the tests in this PR with that change? If I remove`CallableVisitor::VisitFieldDecl()`, I get 5 errors for missing diagnostics still for all the field initializer test cases in `clang/test/SemaCXX/warn-unsafe-buffer-usage-function-attr.cpp`

I will try out `TraverseCXXDefaultInitExpr()` and see what I can see.

danakj replied:
Well, regardless of the value of `MatchDescendantVisitor::shouldVisitImplicitCode()`, there is no call to a `MatchDescendantVisitorTraverseCXXDefaultInitExpr()` occurring in this test:

```cc
// RUN: %clang_cc1 -std=c++20 -Wunsafe-buffer-usage \
// RUN:            -fsafe-buffer-usage-suggestions -verify %s

class UnsafeMembers {
public:
    [[clang::unsafe_buffer_usage]]
    UnsafeMembers(int) {}
};

struct HoldsUnsafeMembers {
    UnsafeMembers FromField{3};  // expected-warning{{function introduces unsafe buffer manipulation}}
};
```

I will keep trying because you saw something that I am missing, but noting my progress here in case it's useful.

danakj replied:
I see. This test has `TraverseCXXDefaultInitExpr` occur:

```cc
// RUN: %clang_cc1 -std=c++20 -Wunsafe-buffer-usage \
// RUN:            -fsafe-buffer-usage-suggestions -verify %s

class UnsafeMembers {
public:
    [[clang::unsafe_buffer_usage]]
    UnsafeMembers(int) {}
};

struct HoldsUnsafeMembers {
    HoldsUnsafeMembers() {}
    
    UnsafeMembers FromField{3};  // expected-warning{{function introduces unsafe buffer manipulation}}
};
```

But if the explicit ctor is omitted, then there is no such `TraverseCXXDefaultInitExpr` and the FieldDecl must be visited, unless there's something else to anchor it on.

haoNoQ replied:
I'm sure I've seen a difference on the tests you mentioned. Let me take another look.

In your new test the entire constructor is missing. The class is an aggregate, it has no constructor: https://godbolt.org/z/aa869TKsf - so I hope you'll see the warning whenever you aggregate-initialize an instance of such class. Which may or may not mean that there's still no direct need to warn at the definition of the field.

danakj replied:
This is the AST in the case without the explicit constructor:
```
  |-CXXRecordDecl <col:1, col:8> col:8 implicit struct HoldsUnsafeMembers
  `-FieldDecl <line:8:5, col:30> col:19 FromField 'UnsafeMembers'
    `-CXXConstructExpr <col:28, col:30> 'UnsafeMembers' 'void (int)' list
      `-IntegerLiteral <col:29> 'int' 3
```

https://godbolt.org/z/5freKecWb

AFAICT we are required to visit FieldDecl to handle the 0-ctor class use case. Thinking through design, maybe we could visit RecordDecl instead, in CallableVisitor, and then provide a TraverseRecordDecl in the MatchDescendantVisitor that would traverse each field initializer? I could try it, I am not sure it would work, but it's not obviously more clear/better to me either.

I added a comment on the aggregate test case mentioning the lack of CXXDefaultInitExpr that it is testing for.

danakj replied:
> In your new test the entire constructor is missing. The class is an aggregate, it has no constructor: https://godbolt.org/z/aa869TKsf - so I hope you'll see the warning whenever you aggregate-initialize an instance of such class. Which may or may not mean that there's still no direct need to warn at the definition of the field.

https://godbolt.org/z/qjqaW3Eze here is an example of the 3 ways to call the aggregate ctor.

```
    |-DeclStmt <line:12:5, col:26>
    | `-VarDecl <col:5, col:24> col:24 m1 'HoldsUnsafeMembers' callinit
    |   `-CXXConstructExpr <col:24> 'HoldsUnsafeMembers' 'void () noexcept(false)'
    |-DeclStmt <line:13:5, col:35>
    | `-VarDecl <col:5, col:34> col:10 m2 'HoldsUnsafeMembers' cinit
    |   `-CXXTemporaryObjectExpr <col:15, col:34> 'HoldsUnsafeMembers' 'void () noexcept(false)' zeroing
    `-DeclStmt <line:14:5, col:35>
      `-VarDecl <col:5, col:34> col:10 m3 'HoldsUnsafeMembers' cinit
        `-CXXFunctionalCastExpr <col:15, col:34> 'HoldsUnsafeMembers' functional cast to HoldsUnsafeMembers <NoOp>
          `-InitListExpr <col:33, col:34> 'HoldsUnsafeMembers'
            `-CXXDefaultInitExpr <col:34> 'UnsafeMembers' has rewritten init
              `-CXXConstructExpr <line:8:28, col:30> 'UnsafeMembers' 'void (int)' list
                `-IntegerLiteral <col:29> 'int' 3
```

We do get a CXXDefaultInitExpr if we construct with `{}` but not otherwise.

haoNoQ replied:
> AFAICT we are required to visit FieldDecl to handle the 0-ctor class use case.

I really want us to momentarily stop being preoccupied with "can" and focus on "should". This isn't a technical problem, this is a UI/UX problem.

There's no constructor. There's no binary code for constructor. The unsafe function is not invoked from any binary code generated from your example. It's the responsibility of the code that aggregate-initializes your aggregate to generate the call to the default initializer, so naturally it doesn't show up just yet.

The class is safe to use whenever the field is initialized explicitly. If they remove the field initializer, that'll break backwards compatibility. Maybe it's ok to keep the class as-is and use the attribute only as a reminder to the callers to call a better constructor.

Ultimately I think you're right and in most of such cases the best solution is to change the class to use a different, safe default initializer for the field. Which presumably exists, given that the attribute was placed on the currently used constructor. I think it's true that more users would complaint about "false negative" in the aggregate definition, or an awkwardly placed / hard-to-read warning at the curly braces (especially if the unsafe initializer is layered into several other layers of aggregates) than complain about false positive under "the class is intended to be safe as long as the field initializer is passed explicitly".

But this is something we should occasionally think about, and treat this as a very very special cornercase. Which doesn't even necessarily require an immediate solution - given that all unsafe code will eventually be flagged regardless. So I think if we go for explicitly supporting fields, we should probably exclude all non-aggregate classes from the callback, and leave an irritated comment explaining why we eventually agreed that it's a good idea 😅

danakj replied:
> There's no constructor. There's no binary code for constructor. The unsafe function is not invoked from any binary code generated from your example.

Ah yes, this helped me figure out what's going on with the 3 different initialization paths.

If you construct the aggregate like `A a;` or `auto a = A();` then we do see a call to a constructor. So where is it? Now it appears in the AST for the class:
```
| |-CXXRecordDecl <col:1, col:8> col:8 implicit struct HoldsUnsafeMembers
| |-FieldDecl <line:8:5, col:30> col:19 FromField 'UnsafeMembers'
| | `-CXXConstructExpr <col:28, col:30> 'UnsafeMembers' 'void (int)' list
| |   `-IntegerLiteral <col:29> 'int' 3
| |-CXXConstructorDecl <line:7:8> col:8 implicit used constexpr HoldsUnsafeMembers 'void () noexcept(false)' inline default
| | |-CXXCtorInitializer Field 0xbc32670 'FromField' 'UnsafeMembers'
| | | `-CXXDefaultInitExpr <col:8> 'UnsafeMembers' has rewritten init
| | |   `-CXXConstructExpr <line:8:28, col:30> 'UnsafeMembers' 'void (int)' list
| | |     `-IntegerLiteral <col:29> 'int' 3
```

If you construct it as `auto a = A{};` then yes there is no ctor as you say, but it does appear at the call site. So okay I agree there's nothing to do for an aggregate that is never used. And maybe I can get TraverseCXXDefaultInitExpr to work with that information.

haoNoQ replied:
Ok here's how I think this should go:
```diff
diff --git a/clang/lib/Analysis/UnsafeBufferUsage.cpp b/clang/lib/Analysis/UnsafeBufferUsage.cpp
index bfbf84a03b2b..cf0194d29f8c 100644
--- a/clang/lib/Analysis/UnsafeBufferUsage.cpp
+++ b/clang/lib/Analysis/UnsafeBufferUsage.cpp
@@ -170,6 +170,12 @@ public:
     return VisitorBase::TraverseCXXTypeidExpr(Node);
   }
 
+  bool TraverseCXXDefaultInitExpr(CXXDefaultInitExpr *Node) {
+    if (!TraverseStmt(Node->getExpr()))
+      return false;
+    return VisitorBase::TraverseCXXDefaultInitExpr(Node);
+  }
+
   bool TraverseStmt(Stmt *Node, DataRecursionQueue *Queue = nullptr) {
     if (!Node)
       return true;
@@ -3343,16 +3349,6 @@ void clang::checkUnsafeBufferUsage(const Decl *D,
         Stmts.push_back(CI->getInit());
       }
     }
-  } else if (const auto *FlD = dyn_cast<FieldDecl>(D)) {
-    // Visit in-class initializers for fields.
-    if (!FlD->hasInClassInitializer())
-      return;
-    Stmts.push_back(FlD->getInClassInitializer());
-    // In a FieldDecl there is no function body, there is only a single
-    // statement, and the suggestions machinery is not set up to handle that
-    // kind of structure yet and would give poor suggestions or likely even hit
-    // asserts.
-    EmitSuggestions = false;
   } else if (isa<BlockDecl>(D) || isa<ObjCMethodDecl>(D)) {
     Stmts.push_back(D->getBody());
   }
diff --git a/clang/lib/Sema/AnalysisBasedWarnings.cpp b/clang/lib/Sema/AnalysisBasedWarnings.cpp
index e356c06ac8c4..b9d0b59ef1db 100644
--- a/clang/lib/Sema/AnalysisBasedWarnings.cpp
+++ b/clang/lib/Sema/AnalysisBasedWarnings.cpp
@@ -2471,38 +2471,6 @@ public:
   CallableVisitor(llvm::function_ref<void(const Decl *)> Callback)
       : Callback(Callback) {}
 
-  bool VisitFieldDecl(FieldDecl *Node) {
-    DeclContext *DeclCtx;
-    if (auto *ID = dyn_cast<ObjCIvarDecl>(Node)) {
-      DeclCtx = cast<DeclContext>(ID->getContainingInterface());
-    } else {
-      RecordDecl *RD = Node->getParent();
-      if (auto *CCRD = dyn_cast<CXXRecordDecl>(RD);
-          CCRD && CCRD->isAggregate()) {
-        // Aggregates have implicitly generated constructors which are not
-        // traversed by our AST visitors at this time. The field initializers in
-        // an aggregate may never be used if the callers always explicitly
-        // initialize them, in which case we do not need to warn on unsafe
-        // buffer usage in the initializer.
-        //
-        // FIXME: We should go through the implicit aggregate initialization
-        // (either an implicit CXXConstructExpr for value-init or an implicit
-        // ListInitExpr for aggregate-init) to determine if a field's default
-        // initializer (CXXDefaultInitExpr) is actually used. If it is, then we
-        // should visit it, while retaining a reference to the caller for
-        // showing the path to the use of the default initializer in the
-        // warning.
-        return true;
-      }
-      DeclCtx = cast<DeclContext>(RD);
-    }
-    if (DeclCtx->isDependentContext())
-      return true; // Not to analyze dependent decl
-    if (Node->hasInClassInitializer())
-      Callback(Node);
-    return true;
-  }
-
   bool VisitFunctionDecl(FunctionDecl *Node) {
     if (cast<DeclContext>(Node)->isDependentContext())
       return true; // Not to analyze dependent decl
diff --git a/clang/test/SemaCXX/warn-unsafe-buffer-usage-function-attr.cpp b/clang/test/SemaCXX/warn-unsafe-buffer-usage-function-attr.cpp
index b9e517a35de7..724d444638b5 100644
--- a/clang/test/SemaCXX/warn-unsafe-buffer-usage-function-attr.cpp
+++ b/clang/test/SemaCXX/warn-unsafe-buffer-usage-function-attr.cpp
@@ -128,7 +128,7 @@ struct HoldsUnsafeMembers {
 
     UnsafeMembers FromCtor;
     UnsafeMembers FromCtor2;
-    UnsafeMembers FromField{3};  // expected-warning{{function introduces unsafe buffer manipulation}}
+    UnsafeMembers FromField{3};  // expected-warning 2{{function introduces unsafe buffer manipulation}}
 };
 
 struct SubclassUnsafeMembers : public UnsafeMembers {
diff --git a/clang/test/SemaCXX/warn-unsafe-buffer-usage-in-container-span-construct.cpp b/clang/test/SemaCXX/warn-unsafe-buffer-usage-in-container-span-construct.cpp
index 945a4e06d6af..22daac83b879 100644
--- a/clang/test/SemaCXX/warn-unsafe-buffer-usage-in-container-span-construct.cpp
+++ b/clang/test/SemaCXX/warn-unsafe-buffer-usage-in-container-span-construct.cpp
@@ -158,9 +158,19 @@ namespace test_flag {
 struct HoldsStdSpan {
   char* Ptr;
   unsigned Size;
-  std::span<char> Span{Ptr, Size};  // expected-warning{{the two-parameter std::span construction is unsafe as it can introduce mismatch between buffer size and the bound information}}
+  std::span<char> Span{Ptr, Size}; // no-warning (this code is unreachable)
 
   HoldsStdSpan(char* P, unsigned S)
       : Span(P, S)  // expected-warning{{the two-parameter std::span construction is unsafe as it can introduce mismatch between buffer size and the bound information}}
   {}
 };
+
+struct HoldsStdSpan2 {
+  char* Ptr;
+  unsigned Size;
+  std::span<char> Span{Ptr, Size}; // expected-warning{{the two-parameter std::span construction is unsafe as it can introduce mismatch between buffer size and the bound information}}
+
+  HoldsStdSpan2(char* P, unsigned S)
+      : Ptr(P), Size(S)
+  {}
+};
```

This passes all tests except two. In one of those we aren't getting a warning but that's ok because the code is obviously dead. I added another test that demonstrates that we do get a warning when it actually matters.

In the other failure we're getting a duplicate warning because we're visiting the same field twice (from two different constructors). This isn't ideal and ultimately we'd need to either get rid of one of them or at least add notes to connect them back to the constructor, but that's not a deal breaker. We may also occasionally emit conflicting fixits because of that (attached to separate warnings) but that's probably still ok; eventually we'll get them deduplicated.

So I basically think that all of this is still better than analyzing field initializers separately; we'd never want to analyze them separately. Also more concise.

The visitor is a bit confusing; the `match(*Node)` invocation should have been put into the `Visit()` methods, not into `Traverse()` methods. Traverse methods are hard to override because their default implementation has tons of logic.

danakj replied:
Thanks, I have (finally) applied that delta, and got the same test results. I also didn't know the `2` trick for duplicate warnings, that's helpful!

danakj commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -3328,3 +3300,63 @@ void clang::checkUnsafeBufferUsage(const Decl *D,
+  for (Stmt *S : Stmts) {
+    auto [FixableGadgets, WarningGadgets, Tracker] =
+        findGadgets(S, D->getASTContext(), Handler, EmitSuggestions);
+    applyGadgets(D, std::move(FixableGadgets), std::move(WarningGadgets),
```
If I understand correctly, we'd still collect a list of gadgets+trackers in the same way, but we'd pass them all at once to applyGadgets() so it could do some kind of global analysis over the full set if it wanted to. Though it will just need to iterate over them for now?

danakj replied:
Or, I guess we'd have lists of gadget, but the difference here is we'd have only a single tracker? So applyGadgets actually just walks the global list of gadgets (it already does this for a list of gadgets) and uses the one tracker... I will try this out.

danakj committed:
applyGadgets once per function and avoid fixits in field initializers

findGadgets() collects all gadgets into a single list, using a
single DeclUseTracker. Then applyGadgets() is called once with the
full list for a method (including for a ctor with initializers).

Handle ObjC interface variables as they return null from getParent().
Instead go through getContainingInterface() to get the DeclContext.

Disable fixits in field initializers as the machinery will not expect
a statement that is not a function body at this time.

danakj commented:
I have rebased (which dropped the previous PR's commits) and applied the requested changes in a second commit. I also fixed the test failures in that second commit.

danakj commented:
bump for review?

danakj committed:
Add comment on AggregateUnsafeMembers

The aggregate test case generates an AST without any ctor and
thus without any ctor initializers, inlcuding no CXXDefaultInitExpr.

danakj committed:
Avoid visiting field initializers in aggregate records

In aggregates, the field initializer may be never invoked in which
case we do not need to visit and warn on it.

FIXMEs are left in place for aggregates where the field initializer
is used, but there's currently no warning.

jkorous-apple commented:
@haoNoQ Do you intend to review the patch?

haoNoQ commented:
I did. The latest comment is https://github.com/llvm/llvm-project/pull/91991#discussion_r1643433741

jkorous-apple commented:
Sorry, I should've asked - do you intend to review the changes that followed your comment?

haoNoQ commented:
Please look at the dates.

danakj commented:
Sorry I just haven't gotten back to making the fixes yet. If someone else wants to apply them and land it I won't be upset, otherwise I will get to this eventually..

ziqingluo-90 commented:
Hi @danakj ,  we are really looking forward to this change!   Will you finish it soon?  Or I can pick it up and land it on behalf of you?

danakj committed:
Avoid FieldDecl visitor

We can visit CXXDefaultInitExprs to find construction through
[[unsafe_buffer_usage]] constructors of field initializers.

danakj commented:
The suggested fixes were applied, thanks for them. It was a bit of work to page all this back in after a while.

ziqingluo-90 commented on clang/lib/Analysis/UnsafeBufferUsage.cpp:
```diff
@@ -3912,3 +3896,56 @@ void clang::checkUnsafeBufferUsage(const Decl *D,
+
+    Stmts.push_back(FD->getBody());
+
+    if (const auto *ID = dyn_cast<CXXConstructorDecl>(D)) {
```
Is this equivalent to override `traverseCXXCtorInitializer` in `MatchDescendantVisitor` above (just like how you override ` TraverseCXXDefaultInitExpr`)?

danakj replied:
I don't think we have that option, because they are not statements: https://github.com/llvm/llvm-project/blob/5545f76dc94e76ef6800823bdd1e107ad2264717/clang/include/clang/Basic/StmtNodes.td#L133-L134

That's why this code has to pull the `getInit()` out of it to find a statement.

ziqingluo-90 reviewed (approved):
LGTM!  Thank you @danakj for the continuous effort!

ZequanWu commented:
Hi, we have found this causing clang to crash with -Wunsafe-buffer-usage: https://godbolt.org/z/nvWxW43fK

ZequanWu commented:
It's actually an existing crash not caused by this one.

ZequanWu commented:
Confirmed that this change causes crash with -Wunsafe-buffer-usage: https://godbolt.org/z/vcv8d6sYr. 

My previously manual-reduced repro in https://github.com/llvm/llvm-project/pull/91991#issuecomment-2477311737 happens to crash before this change.

ziqingluo-90 commented:
@ZequanWu Thanks for the bug-catching!  I will look into it!

zmodem commented:
@ZequanWu Based on the discussion on #116286 it sounds like the crashes are now fixed? Can you double check, and re-land this if so?

ZequanWu commented:
Verified that it no longer crashes.
